lib/python/rosie/ws.py

```diff
@@ -189,33 +234,173 @@ def _render(self, all_revs=0, data=None, filters=None, s=None):
+        data = self.dao.query(filters, all_revs)
+        if self.get_query_argument("format", default=None) == "json":
+            self.write(json.dumps(data))
+        else:
+            self._render(all_revs, data, filters=filters)
+
+    @staticmethod
+    def _query_parse_string(q_str):
+        """Split a query filter string into component parts."""
+        conjunction, tail = q_str.split(" ", 1)
+        if conjunction == "or" or conjunction == "and":
+            q_str = tail
+        else:
+            conjunction = "and"
+        filt = [conjunction]
+        if all(s == "(" for s in q_str.split(" ", 1)[0]):
+            start_group, q_str = q_str.split(" ", 1)
+            filt.append(start_group)
+        key, operator, value = q_str.split(" ", 2)
+        filt.extend([key, operator])
+        last_groups = value.rsplit(" ", 1)
+        if (len(last_groups) > 1 and last_groups[1] and
+                all([s == ")" for s in last_groups[1]])):
+            filt.extend(last_groups)
+        else:
+            filt.extend([value])
+        return filt
+
+
+def _log_server_status(application, host, port):
+    """ Log a brief status, including process ID, for an application server."""
+    log_root = os.path.expanduser(LOG_ROOT_TMPL % {
+        "ns": application.NAMESPACE,
+        "util": application.UTIL,
+        "host": host,
+        "port": port})
+    log_status = log_root + ".status"
+    if not os.path.isdir(os.path.dirname(log_root)):
+        os.makedirs(os.path.dirname(log_root))
+    with open(log_status, "w") as handle:
+        handle.write("host=%s\n" % host)
+        handle.write("port=%d\n" % port)
+        handle.write("pid=%d\n" % int(application.get_app_pid()))
+    return log_status
+
+
+def _get_server_status(application, host, port):
+    """Return a dictionary containing a brief application server status."""
+    ret = {}
+    log_root_glob = os.path.expanduser(LOG_ROOT_TMPL % {
+        "ns": application.NAMESPACE,
+        "util": application.UTIL,
+        "host": "*",
+        "port": "*"})
+    for filename in glob(log_root_glob):
+        try:
+            for line in open(filename):
+                key, value = line.strip().split("=", 1)
+                ret[key] = value
+            break
+        except (IOError, ValueError):
+            pass
+    return ret
+
+
+def main():
+    arg_msg_comp = "ad-hoc web service server"
+    # No value req for Boolean options e.g. --start equivalent to --start=True.
+    define("start", type=bool, default=False,
+           help="start %s (on specified port, else on port %s)" % (
+               arg_msg_comp, DEFAULT_PORT))
+    define("port", type=int, default=DEFAULT_PORT, help="port to listen on")
+    define("stop", type=bool, default=False, help="stop %s" % arg_msg_comp)
+    define("non-interactive", type=bool, default=False,
+           help="to stop %s w/o prompting" % arg_msg_comp)
+    parse_command_line()
```

Comment:
I changed to use the Tornado command-line parsing module thinking it would be cleaner & reduce the line count for management of the ad-hoc server, but considering it again & in light of your reasoning I can see that is not really true. I'll revert the CLI & adapt the PR accordingly. Thanks.